lib/spack/spack/util/module_cmd.py

```diff
@@ -162,6 +158,41 @@ def path_from_modules(modules):
         # that we give preference to the last module to be loaded
         # for packages requiring to load multiple modules in sequence
         best_choice = candidate_path or best_choice
+
+    # if we found something return it
+    return best_choice
+
+
+def path_from_modules(modules, hint_name=None):
+    """Inspect a list of TCL modules for entries that indicate the absolute
+    path at which the library supported by said module can be found.
+
+    Args:
+        modules (list): module files to be loaded to get an external package
+        hint_name (str): optional hint giving a name that likely maps
+                         is a better choice than choosing the last
+                         possible path.
+                         A good choice is the package name.
+                         E.g., if operating on an external mvapich,
+                         then give priority to modules with '.*mvapich.*'
+                         in their name. From these, always choose the last
+
+    Returns:
+        Guess of the prefix path where the package
+    """
+    modules = list(modules)  # defensive copy for generators
```

Comment:
I agree defensive copying. Maybe we should also be defensive on strings passed in, since:
```
Spack version 0.16.1
Python 3.6.9, Linux x86_64
>>> modules = 'gcc/9.3.0'
>>> modules = list(modules)
>>> modules
['g', 'c', 'c', '/', '9', '.', '3', '.', '0']
>>> 

```

Reply:
@alalazo -

Can propose a good string test (before the list copy).

E.g.,
https://stackoverflow.com/questions/4843173/how-to-check-if-type-of-a-variable-is-string

Seems to say that between python 2 and 3 you can get bitten by Unicode nonsense if you aren't careful

Reply:
@jjellio Depending on whether or not we want to accept strings, I would do either:
```python
# Normalize input
if isinstance(modules, six.string_types):
    modules = [modules]

modules = list(modules)
```
or
```python
assert not isinstance(modules, six.string_types), "modules cannot be a string"
modules = list(modules)
```

Reply:
It is safe to assume you have `six` ?

I committed a py 2/3 compliant check, (see current state of the file)

The gist of the code is the same (dump the string in a list) - the critique is on how to detect the string types. 

The package-free method seems like checking if python >= 3, then define `unicode=str`. Then test for `str` or `unicode` (py3 will end up with redundant checks - py2 will have two type checks)

Either way I did push that.  Can use six if that's spack's preferred way.

Reply:
> It is safe to assume you have six ?

Yeah, we vendor `six` along with a few other pure python dependencies.

Reply:
I can amend to use `six` then.

Reply:
Pushed use of `six`